DSDEGP-1770: replace gsutil with java api

pshapiro4broad opened the pull request:
DSDEGP-1770: replace gsutil shell subprocess calls with native java API

- remove unused APIs and tests from IoUtil
- fix error reporting code so exception types, stack crawls aren't lost
- update readme to match current integration test names

----

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-1770: replace gsutil with google storage APIs

- instead of using a subprocess to run gsutil, use the native java APIs
- remove unused APIs and tests from IoUtil
- fix error reporting code so exception types, stack crawls aren't lost

pshapiro4broad committed:
DSDEGP-1770: update readme to match current integration test names

pshapiro4broad committed:
DSDEGP-1770: update tests to match code

tlangs commented on clio-client/src/test/scala/org/broadinstitute/clio/client/util/IoUtilSpec.scala:
```diff
@@ -1,101 +1,37 @@
-      IoUtil.readMetadata(f.write(contents).uri) should be(contents)
-    }
-  }
+    override def readGoogleObjectData(location: URI): String = { " " }
```
The braces here are unnecessary

tlangs commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala:
```diff
@@ -54,163 +52,94 @@ trait IoUtil {
-  }
+  def deleteGoogleObject(path: URI): Unit
 
-  def getMd5HashOfGoogleObject(path: URI): Option[Symbol] = {
```
Other projects depend on md5 hashes.

pshapiro4broad replied:
There is an API to get it, so it's easy to implement if we need it. What is using it? I looked at backfill and workflow_tests and didn't see it used there. It's not used in clio. The backfill does override this method in `NioUtil` but it's not called, as far as I can tell.

tlangs replied:
I thought the backfill uses it when backfilling wgs ubams. maybe we stopped doing that?

danxmoran replied:
There's a separate API that gets existence, size, and md5 all at once.

Either way I think it's OK to remove things here that the main clio code doesn't use, as long as the backfill code can extend `IoUtil` with new methods.

pshapiro4broad replied:
The code in backfill uses `getGoogleObjectInfo()`, which is implemented using the storage API, not gsutil or the NIO API. That seems OK although there are some small overlaps with the code here and the code in backfill; we can decide what to do about backfill in a later checkin. Once this settles down I'll see about updating backfill.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala:
```diff
@@ -54,163 +52,94 @@ trait IoUtil {
-    def ls(path: String): Seq[String] = {
-      runGsUtilAndGetStdout(Seq("ls", path)).split("\n")
-    }
+class GsUtil(credentials: Credentials) extends IoUtil {
```
Would still be a fan of what we talked about earlier:
1. Fold `GsUtil` into `IoUtil`
2. Pass a `Storage` instance into the `IoUtil` constructor
3. Move the existing `Credentials -> Storage` logic into an `apply` method on `IoUtil`

Once `Storage` is exposed in the constructor we can mock it out for unit tests, and get that coverage % up. The `apply` logic would be the only bit leaned on by the integration tests.

pshapiro4broad replied:
OK now that it's working I'll give this a try.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala:
```diff
@@ -54,163 +52,94 @@ trait IoUtil {
-      runGsUtilAndGetStdout(Seq("stat", path))
+  override def deleteGoogleObject(path: URI): Unit = {
+    if (!getBlob(path).delete()) {
+      throw new RuntimeException(s"Cannot delete '$path' because it does not exist.")
```
Maybe an `IllegalArgumentException` instead, to be more specific?

pshapiro4broad replied:
OK. Actually I'm unsure how this can occur in practice since `getBlob()` seems to return null if an object doesn't exit.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala:
```diff
@@ -54,163 +52,94 @@ trait IoUtil {
+        } finally {
+          channel.close()
+        }
+      case _ =>
```
What value could `_` be here? If it's `null`, then `getBlob` should be changed to return an `Option`.

pshapiro4broad replied:
Yeah, I thought about that but didn't want to change `read` to return `Option[String]`, etc. I'll look at it again.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala:
```diff
@@ -54,163 +52,94 @@ trait IoUtil {
+  override def googleObjectExists(path: URI): Boolean = {
+    getBlob(path) match {
+      case blob: Blob => blob.exists()
+      case _          => false
```
Same comment about `getBlob`'s return type. With an `Option` this'd be pretty concise.

danxmoran commented on clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeleteExecutorSpec.scala:
```diff
@@ -166,7 +167,9 @@ class DeleteExecutorSpec extends BaseClientSpec with AsyncMockFactory {
       paths.foreach { uri =>
-        ex.getMessage should include(uri.toString)
+        val sw = new StringWriter
+        ex.printStackTrace(new PrintWriter(sw))
```
Could move the printing out of the `foreach`

pshapiro4broad replied:
good point, `ex` doesn't change inside the  `foreach`.

danxmoran commented on clio-client/src/test/scala/org/broadinstitute/clio/client/util/IoUtilSpec.scala:
```diff
@@ -104,44 +40,56 @@ class IoUtilSpec extends BaseClientSpec with AsyncMockFactory {
+      _ <- recoverToExceptionIf[IOException](stream.runWith(Sink.ignore)).map { ex =>
+        urisToFail.foreach { uri =>
+          val sw = new StringWriter
+          ex.printStackTrace(new PrintWriter(sw))
```
Could move the printing out of the `foreach`

danxmoran commented on clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/BaseIntegrationSpec.scala:
```diff
@@ -59,10 +60,25 @@ abstract class BaseIntegrationSpec(clioDescription: String)
     */
   def clioWebClient: ClioWebClient
 
+  private lazy val googleCredential: Credentials =
```
An `IoUtil.apply` taking whatever type `serviceAccount` is would save us from needing to duplicate this logic.

pshapiro4broad replied:
I think this is the only code that does this, at least before calling `IoUtil()`.

pshapiro4broad committed:
DSDEGP-1770: remove unnecessary { }

danxmoran reviewed (commented):
Oops forgot to submit the review

pshapiro4broad committed:
DSDEGP-1770: lift loop invariant

pshapiro4broad committed:
DSDEGP-1770: code review feedback

- fold GsUtil into IoUtil
- replace null with Option
- increase test coverage of new code by using in-memory google storage object

pshapiro4broad committed:
DSDEGP-1770: code review feedback, minor changes

pshapiro4broad committed:
DSDEGP-1770: replace `.head` call with `match`

pshapiro4broad committed:
DSDEGP-1770: scalafmt

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps_DSDEGP-1770_replace_gsutil_with_java_api

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala:
```diff
@@ -47,7 +47,7 @@ class AddExecutor[CI <: ClioIndex](addCommand: AddCommand[CI]) extends Executor
     Source
       .single(location)
-      .map(ioUtil.readMetadata)
+      .map(ioUtil.readFileData)
```
This change busted the expectations for the `IoUtil` mock in the unit tests.

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/sbt/Dependencies.scala:
```diff
@@ -35,6 +35,7 @@ object Dependencies {
   private val TestContainersScalaVersion = "0.16.0"
   private val TypesafeConfigVersion = "1.3.3"
   private val VaultJavaDriverVersion = "3.1.0"
+  private val GoogleCloudStorageVersion = "1.14.0"
```
Pls put in ABC order

danxmoran reviewed (approved):
Final nitpicks, LGTM once unit tests are fixed up

pshapiro4broad committed:
DSDEGP-1770: allow subclasses in backfill to call `getBlob`, `requireBlob`

pshapiro4broad committed:
DSDEGP-1770: fix broken tests, alpha sort dependencies

pshapiro4broad committed:
DSDEGP-1770: replace `try` with new library `scala-arm`, update match to use same syntax as other matches
